nextcloud/server
# Fetch a new request token as soon as the browser becomes online

Fixes https://github.com/nextcloud/server/issues/16809

Debugging a few Sentry tickets I saw evidence of HTTP 412's -> CSRF check failing. Usually those are unexpected as we periodically fetch the current token, but if the client machine is paused or the browser goes offline, the CSRF token might not be fetched when the connection returns. Thus I added event handlers that can pause the background requests when there is no network connectivity and send off a request for the new token as soon as there is connectivity.

![Bildschirmfoto von 2020-01-09 14-14-39](https://user-images.githubusercontent.com/1374172/72070799-a75f5400-32ea-11ea-8f04-6ddd272e85f6.png)

How to test
* Open Nextcloud
* Switch browser to offline mode (File -> Work offline on FF, dev tools for Chrome)
* See detection of offline mode
* Restart web server (to force new token)
* Disable offline mode
* See detection and new token in console

As this is something apps might have interest in, e.g. when to resume operation after offline as you have a dependency on the CSRF token, I added events emitted to the event bus where apps can listen to both the browser going offline as well as coming back online.

Do we want this in 18 already? Then we can backport.
